extensions/amp-access/0.1/amp-access.js

```diff
@@ -393,141 +303,31 @@ export class AccessService {
-  collectUrlVars_(url, useAuthData) {
-    return this.prepareUrlVars_(useAuthData).then(vars => {
-      return this.urlReplacements_.collectVars(url, vars);
-    });
-  }
-
-  /**
-   * @param {boolean} useAuthData Allows `AUTH(field)` URL var substitutions.
-   * @return {!Promise<!Object<string, *>>}
-   * @private
-   */
-  prepareUrlVars_(useAuthData) {
-    return this.getReaderId_().then(readerId => {
-      const vars = {
-        'READER_ID': readerId,
-        'ACCESS_READER_ID': readerId,  // A synonym.
-        'ACCESS_TOKEN': () => this.signIn_.getAccessTokenPassive(),
-      };
-      if (useAuthData) {
-        vars['AUTHDATA'] = field => {
-          if (this.authResponse_) {
-            return getValueForExpr(this.authResponse_, field);
-          }
-          return undefined;
-        };
-      }
-      return vars;
-    });
-  }
-
   /**
    * Returns the promise that resolves when all authorization work has
    * completed, including authorization endpoint call and UI update.
-   * Note that this promise never fails.
+   * Note that this promise never fails. UI is updated as each response arrives.
    * @param {boolean=} opt_disableFallback
    * @return {!Promise}
    * @private
    */
   runAuthorization_(opt_disableFallback) {
-    if (!this.adapter_.isAuthorizationEnabled()) {
-      dev().fine(TAG, 'Ignore authorization for type=', this.type_);
-      this.firstAuthorizationResolver_();
-      return Promise.resolve();
-    }
-
-    this.toggleTopClass_('amp-access-loading', true);
-    const startPromise = this.viewer_.whenFirstVisible();
-    const responsePromise = startPromise.then(() => {
-      return this.adapter_.authorize();
-    }).catch(error => {
-      this.analyticsEvent_('access-authorization-failed');
-      if (this.authorizationFallbackResponse_ && !opt_disableFallback) {
-        // Use fallback.
-        user().error(TAG, 'Authorization failed: ', error);
-        return this.authorizationFallbackResponse_;
-      } else {
-        // Rethrow the error, it will be processed in the bottom `catch`.
-        throw error;
-      }
-    });
-    const promise = responsePromise.then(response => {
-      dev().fine(TAG, 'Authorization response: ', response);
-      this.setAuthResponse_(response);
-      this.toggleTopClass_('amp-access-loading', false);
-      this.toggleTopClass_('amp-access-error', false);
-      this.buildLoginUrls_();
-      return this.ampdoc.whenReady().then(
-          () => this.applyAuthorization_(response));
-    }).catch(error => {
-      user().error(TAG, 'Authorization failed: ', error);
-      this.toggleTopClass_('amp-access-loading', false);
-      this.toggleTopClass_('amp-access-error', true);
-    });
-    // The "first" promise must always succeed first.
-    this.lastAuthorizationPromise_ = Promise.all(
-        [this.firstAuthorizationPromise_, promise]);
+    const promises = this.sources_.map(source =>
+      source.runAuthorization()
+          .then(response =>
```

Comment:
Could you pls add some `{}` here? In this particular case shortcuts are hard to read.

Reply:
this was making my head hurt, too. factored out a little helper method